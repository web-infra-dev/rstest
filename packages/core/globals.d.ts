declare global {
  const test: typeof import('@rstest/core')['test'];
  const describe: typeof import('@rstest/core')['describe'];
  const it: typeof import('@rstest/core')['it'];
  const expect: typeof import('@rstest/core')['expect'];
}
export {};
