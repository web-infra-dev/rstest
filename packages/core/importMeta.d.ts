interface ImportMeta {
  readonly rstest?: typeof import('@rstest/core');
}
