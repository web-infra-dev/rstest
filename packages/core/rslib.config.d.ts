import { RslibConfig } from '@rslib/core';
declare const config: RslibConfig;
export default config;
