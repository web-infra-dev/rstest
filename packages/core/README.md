<picture>
  <img alt="Rstest Banner" src="https://assets.rspack.dev/rstest/rstest-banner.png">
</picture>

# Rstest

Rstest is a testing framework powered by Rspack.

## ðŸ“– License

Rstest is licensed under the [MIT License](https://github.com/web-infra-dev/rstest/blob/main/LICENSE).
