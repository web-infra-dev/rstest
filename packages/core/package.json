{
  "name": "@rstest/core",
  "version": "0.0.0",
  "description": "The Rsbuild-based test tool.",
  "bugs": {
    "url": "https://github.com/web-infra-dev/rstest/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rstest",
    "directory": "packages/core"
  },
  "license": "MIT",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist-types/index.d.ts",
  "bin": {
    "rstest": "./bin/rstest.js"
  },
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "default": "./dist/index.js"
    },
    "./globals": {
      "types": "./globals.d.ts"
    }
  },
  "files": [
    "bin",
    "dist",
    "dist-types",
    "compiled",
    "types.d.ts"
  ],
  "scripts": {
    "build": "rslib build",
    "dev": "rslib build --watch",
    "test": "npx rstest run --globals"
  },
  "dependencies": {
    "@rsbuild/core": "^1.2.19",
    "birpc": "2.2.0",
    "chai": "^5.2.0",
    "std-env": "^3.8.1",
    "tinypool": "^1.0.2"
  },
  "devDependencies": {
    "@babel/code-frame": "^7.26.2",
    "@jridgewell/trace-mapping": "0.3.25",
    "@rslib/core": "0.5.5",
    "@rstest/tsconfig": "workspace:*",
    "@types/babel__code-frame": "^7.0.6",
    "@vitest/expect": "^3.0.9",
    "cac": "^6.7.14",
    "picocolors": "^1.1.1",
    "rslog": "^1.2.3",
    "stacktrace-parser": "0.1.11",
    "tinyglobby": "^0.2.12"
  },
  "engines": {
    "node": ">=16.7.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/"
  }
}
