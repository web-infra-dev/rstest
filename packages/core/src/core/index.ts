import type { RstestCommand, RstestConfig, RstestInstance } from '../types';
import { createContext } from './context';

export function createRstest(
  config: RstestConfig,
  command: RstestCommand,
  fileFilters: string[],
): RstestInstance {
  const context = createContext({ cwd: process.cwd(), command }, config);

  const runTests = async (): Promise<void> => {
    const { runTests } = await import('./runTests');
    await runTests(context, fileFilters);
  };

  return {
    runTests,
  };
}
