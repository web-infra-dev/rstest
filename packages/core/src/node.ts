export { defineConfig } from './public';
export { createRstest } from './core';
export type { RstestConfig, Reporter, RstestCommand } from './types';
