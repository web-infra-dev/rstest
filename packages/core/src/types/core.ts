import type { SnapshotManager } from '@vitest/snapshot/manager';
import type { NormalizedConfig, RstestConfig } from './config';
import type { Reporter } from './reporter';

export type RstestCommand = 'watch' | 'run';

export type RstestContext = {
  /** The Rstest core version. */
  version: string;
  /** The root path of current project. */
  rootPath: string;
  /** The original Rstest config passed from the createRstest method. */
  originalConfig: Readonly<RstestConfig>;
  /** The normalized Rstest config. */
  normalizedConfig: NormalizedConfig;
  /**
   * The command type.
   *
   * - dev: `rstest dev`
   * - run: `rstest run`
   */
  command: RstestCommand;
  reporters: Reporter[];
  snapshotManager: SnapshotManager;
};

export type RstestInstance = {
  runTests: () => Promise<void>;
};
