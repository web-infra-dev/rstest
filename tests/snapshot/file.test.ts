import fs from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from '@rstest/core';
import { runRstestCli, waitFile } from '../scripts';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('test snapshot', () => {
  it('should generator snapshot file correctly with -u', async () => {
    const snapshotFilePath = join(
      __dirname,
      'fixtures/__snapshots__/index.test.ts.snap',
    );

    fs.rmSync(snapshotFilePath, {
      force: true,
    });

    await runRstestCli({
      command: 'rstest',
      args: ['run', '-u', 'fixtures/index.test.ts'],
      options: {
        nodeOptions: {
          cwd: __dirname,
        },
      },
    });

    await waitFile(snapshotFilePath, 3000);
    expect(fs.existsSync(snapshotFilePath)).toBeTruthy();

    const content = fs.readFileSync(snapshotFilePath, 'utf-8');

    // should generator snapshot name correctly
    expect(content).toContain(
      '[`test snapshot > test snapshot generate 1`] = `"hello world"`',
    );

    fs.rmSync(snapshotFilePath);
  });

  describe('test snapshot file state', () => {
    it('test toMatchSnapshot API', () => {
      expect('hello world').toMatchSnapshot();
      expect('hello Rstest').toMatchSnapshot();
    });

    it('test toMatchSnapshot API', () => {
      // test repeat test case name
      expect('hello world 1').toMatchSnapshot();
    });

    it('test toMatchSnapshot API - 1', () => {
      expect('hello world - 1').toMatchSnapshot();
    });

    it('test toMatchSnapshot name', () => {
      expect('hi').toMatchSnapshot('say hi');
    });
  });

  it('test toMatchFileSnapshot correctly', async () => {
    expect('hello world').toMatchFileSnapshot('__snapshots__/file.output.txt');
  });

  it('should failed when snapshot unmatched', async () => {
    const { cli } = await runRstestCli({
      command: 'rstest',
      args: ['run', 'fixtures/fail.test.ts'],
      options: {
        nodeOptions: {
          cwd: __dirname,
        },
      },
    });

    await cli.exec;
    expect(cli.exec.process?.exitCode).toBe(1);

    const logs = cli.stdout.split('\n').filter(Boolean);

    expect(logs.find((log) => log.includes('Snapshots 1 failed'))).toBeTruthy();
  });
});
