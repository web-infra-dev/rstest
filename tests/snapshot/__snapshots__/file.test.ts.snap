// Snapshot v1

exports[`test snapshot > test snapshot file state > test toMatchSnapshot API - 1 1`] = `"hello world - 1"`;

exports[`test snapshot > test snapshot file state > test toMatchSnapshot API 1`] = `"hello world"`;

exports[`test snapshot > test snapshot file state > test toMatchSnapshot API 2`] = `"hello Rstest"`;

exports[`test snapshot > test snapshot file state > test toMatchSnapshot API 3`] = `"hello world 1"`;

exports[`test snapshot > test snapshot file state > test toMatchSnapshot name > say hi 1`] = `"hi"`;
